File: pm_core/src/infrastructure/store/mapper/work_item_mapper.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package mapper

import (
	"github.com/serp/pm-core/src/core/domain/entity"
	"github.com/serp/pm-core/src/infrastructure/store/model"
)

type WorkItemMapper struct{}

func NewWorkItemMapper() *WorkItemMapper {
	return &WorkItemMapper{}
}

func (m *WorkItemMapper) ToEntity(mdl *model.WorkItemModel) *entity.WorkItemEntity {
	if mdl == nil {
		return nil
	}

	return &entity.WorkItemEntity{
		BaseEntity: entity.BaseEntity{
			ID:        mdl.ID,
			CreatedAt: mdl.CreatedAt.UnixMilli(),
			UpdatedAt: mdl.UpdatedAt.UnixMilli(),
		},
		ProjectID:           mdl.ProjectID,
		ItemNumber:          mdl.ItemNumber,
		Title:               mdl.Title,
		Description:         mdl.Description,
		Type:                mdl.Type,
		Status:              mdl.Status,
		Priority:            mdl.Priority,
		ParentID:            mdl.ParentID,
		SprintID:            mdl.SprintID,
		MilestoneID:         mdl.MilestoneID,
		BoardColumnID:       mdl.BoardColumnID,
		ReporterID:          mdl.ReporterID,
		StoryPoints:         mdl.StoryPoints,
		EstimatedHours:      mdl.EstimatedHours,
		StartDateMs:         mdl.StartDateMs,
		DueDateMs:           mdl.DueDateMs,
		ActiveStatus:        mdl.ActiveStatus,
		Position:            mdl.Position,
		HasChildren:         mdl.HasChildren,
		ChildCount:          mdl.ChildCount,
		CompletedChildCount: mdl.CompletedChildCount,
		CommentCount:        mdl.CommentCount,
	}
}

func (m *WorkItemMapper) ToModel(e *entity.WorkItemEntity) *model.WorkItemModel {
	if e == nil {
		return nil
	}

	return &model.WorkItemModel{
		BaseModel: model.BaseModel{
			ID: e.ID,
		},
		ProjectID:           e.ProjectID,
		ItemNumber:          e.ItemNumber,
		Title:               e.Title,
		Description:         e.Description,
		Type:                e.Type,
		Status:              e.Status,
		Priority:            e.Priority,
		ParentID:            e.ParentID,
		SprintID:            e.SprintID,
		MilestoneID:         e.MilestoneID,
		BoardColumnID:       e.BoardColumnID,
		ReporterID:          e.ReporterID,
		StoryPoints:         e.StoryPoints,
		EstimatedHours:      e.EstimatedHours,
		StartDateMs:         e.StartDateMs,
		DueDateMs:           e.DueDateMs,
		ActiveStatus:        e.ActiveStatus,
		Position:            e.Position,
		HasChildren:         e.HasChildren,
		ChildCount:          e.ChildCount,
		CompletedChildCount: e.CompletedChildCount,
		CommentCount:        e.CommentCount,
	}
}

func (m *WorkItemMapper) ToEntities(models []*model.WorkItemModel) []*entity.WorkItemEntity {
	if models == nil {
		return nil
	}

	entities := make([]*entity.WorkItemEntity, 0, len(models))
	for _, mdl := range models {
		if e := m.ToEntity(mdl); e != nil {
			entities = append(entities, e)
		}
	}

	return entities
}

File: pm_core/src/main/java/serp/project/pmcore/infrastructure/store/mapper/WorkItemMapper.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.pmcore.infrastructure.store.mapper;

import org.springframework.stereotype.Component;
import serp.project.pmcore.core.domain.entity.WorkItemEntity;
import serp.project.pmcore.infrastructure.store.model.WorkItemModel;

import java.util.Collections;
import java.util.List;

@Component
public class WorkItemMapper extends BaseMapper {
    public WorkItemModel toModel(WorkItemEntity entity) {
        if (entity == null) {
            return null;
        }
        return WorkItemModel.builder()
                .id(entity.getId())
                .tenantId(entity.getTenantId())
                .projectId(entity.getProjectId())
                .issueTypeId(entity.getIssueTypeId())
                .issueNo(entity.getIssueNo())
                .key(entity.getKey())
                .summary(entity.getSummary())
                .description(entity.getDescription())
                .statusId(entity.getStatusId())
                .priorityId(entity.getPriorityId())
                .assigneeId(entity.getAssigneeId())
                .reporterId(entity.getReporterId())
                .parentId(entity.getParentId())
                .resolutionId(entity.getResolutionId())
                .dueDate(longToLocalDateTime(entity.getDueDate()))
                .rank(entity.getRank())
                .timeOriginalEstimate(entity.getTimeOriginalEstimate())
                .timeRemainingEstimate(entity.getTimeRemainingEstimate())
                .timeSpent(entity.getTimeSpent())
                .createdAt(longToLocalDateTime(entity.getCreatedAt()))
                .createdBy(entity.getCreatedBy())
                .updatedAt(longToLocalDateTime(entity.getUpdatedAt()))
                .updatedBy(entity.getUpdatedBy())
                .build();
    }

    public WorkItemEntity toEntity(WorkItemModel model) {
        if (model == null) {
            return null;
        }
        return WorkItemEntity.builder()
                .id(model.getId())
                .tenantId(model.getTenantId())
                .projectId(model.getProjectId())
                .issueTypeId(model.getIssueTypeId())
                .issueNo(model.getIssueNo())
                .key(model.getKey())
                .summary(model.getSummary())
                .description(model.getDescription())
                .statusId(model.getStatusId())
                .priorityId(model.getPriorityId())
                .assigneeId(model.getAssigneeId())
                .reporterId(model.getReporterId())
                .parentId(model.getParentId())
                .resolutionId(model.getResolutionId())
                .dueDate(localDateTimeToLong(model.getDueDate()))
                .rank(model.getRank())
                .timeOriginalEstimate(model.getTimeOriginalEstimate())
                .timeRemainingEstimate(model.getTimeRemainingEstimate())
                .timeSpent(model.getTimeSpent())
                .createdAt(localDateTimeToLong(model.getCreatedAt()))
                .createdBy(model.getCreatedBy())
                .updatedAt(localDateTimeToLong(model.getUpdatedAt()))
                .updatedBy(model.getUpdatedBy())
                .build();
    }

    public List<WorkItemEntity> toEntities(List<WorkItemModel> models) {
        if (models == null || models.isEmpty()) {
            return Collections.emptyList();
        }
        return models.stream().map(this::toEntity).toList();
    }
}
